File: src/main/java/org/cubexmc/metro/train/TrainStateMachine.java
package org.cubexmc.metro.train;

/**
 * Owns state transitions and transition diagnostics for a train session.
 */
public class TrainStateMachine {

    private final TrainSession session;

    public TrainStateMachine(TrainSession session) {
        this.session = session;
    }

    public TrainMovementTask.TrainState transitionTo(TrainMovementTask.TrainState nextState, String detail) {
        TrainMovementTask.TrainState previousState = session.getState();
        session.setState(nextState);
        session.debug("State transition " + previousState + " -> " + nextState
                + " for passenger=" + session.safePassengerName()
                + ", currentStop=" + session.getCurrentStopId()
                + ", targetStop=" + session.getTargetStopId()
                + (detail == null || detail.isBlank() ? "" : ", " + detail));
        return previousState;
    }
}

File: src/main/java/org/cubexmc/metro/train/TrainStateMachine.kt
package org.cubexmc.metro.train

/**
 * Owns state transitions and transition diagnostics for a train session.
 */
class TrainStateMachine(private val session: TrainSession) {
    fun transitionTo(nextState: TrainMovementTask.TrainState, detail: String?): TrainMovementTask.TrainState {
        val previousState = session.state
        session.state = nextState
        session.debug(
            "State transition $previousState -> $nextState" +
                " for passenger=${session.safePassengerName()}" +
                ", currentStop=${session.currentStopId}" +
                ", targetStop=${session.targetStopId}" +
                if (detail == null || detail.isBlank()) "" else ", $detail",
        )
        return previousState
    }
}
